File: bin/424.dart
import 'dart:math';

void main(List<String> args) {
  print(characterReplacement('ABAB', 2));
  print(characterReplacement('AABABBA', 1));
}

int characterReplacement(String s, int k) {
  var res = 0;

  // 初始化数组是方式
  var counter = List.generate(26, (index) => 0);

  var l = 0;
  // dart的字符串是utf16
  var a = 'A'.codeUnitAt(0);
  for (var r = 0; r < s.length; r++) {
    var idx = s.codeUnitAt(r) - a;
    counter[idx]++;
    var len = r - l + 1;
    if (len - maxCounter(counter) <= k) {
      res = max(res, len);
    }
    while (r - l + 1 - maxCounter(counter) > k) {
      var idx = s.codeUnitAt(l) - a;
      counter[idx]--;
      l++;
    }
  }

  return res;
}

int maxCounter(List<int> counter) {
  var res = counter[0];
  for (var v in counter) {
    if (res < v) {
      res = v;
    }
  }
  return res;
}

File: src/bin/424.rs
fn main() {
    let inputs = vec![("ABAB", 2), ("AABABBA", 1)];

    for (s, k) in inputs {
        let r = Solution::character_replacement(s.into(), k);
        println!(">>> {r}");
    }
}

struct Solution;

impl Solution {
    pub fn character_replacement(s: String, k: i32) -> i32 {
        let bs = s.as_bytes();
        let k = k as usize;
        let mut counter = [0usize; 26];

        let mut l = 0;
        let mut res = 0;

        for (r, c) in bs.iter().enumerate() {
            let idx = (c - b'A') as usize;
            counter[idx] += 1;
            if r - l + 1 - counter.iter().max().unwrap() <= k {
                res = res.max(r - l + 1);
            }
            while r - l + 1 - counter.iter().max().unwrap() > k {
                let c = bs[l];
                let idx = (c - b'A') as usize;
                counter[idx] -= 1;
                l += 1;
            }
        }

        return res as i32;
    }
}
